### ISSUE TITLE
Creating a new folder in a folder provided by a `FileSystemProvider` with a parameter in the URI causes an unnecessary reloading of the root folder

### ISSUE THREAD

**Issue by isc-bsaviano** [microsoft/vscode#139383]
Issue Type: <b>Bug</b>

Steps to reproduce:

1. Clone the sources for the MemFS sample extension.
2. Add this logging line at the beginning of `readDirectory()` on line 45:
```javascript
console.log(`Reading directory ${uri.path}`);
```
3. Add this logging line at the beginning of `createDirectory()` on line 143:
```javascript
console.log(`Creating directory ${uri.path}`);
```
4. Launch the MemFS extension in debug mode.
5. Create the following workspace and open it:
```json
{
  "folders": [
    {
      "name": "MemFS - Sample",
      "uri": "memfs:/"
    }
  ],
  "settings": {}
}
```
6. Create new folder `aaa` in `MemFS - Sample`. You should see the following text logged to your `Debug Console`:
```
Reading directory /
Creating directory /aaa
Reading directory /aaa
```
7. Edit the workpsace so it now looks like this:
```json
{
  "folders": [
    {
      "name": "MemFS - Sample",
      "uri": "memfs:/?notaparam=1"
    }
  ],
  "settings": {}
}
```
8. Create new folder `aaa` in `MemFS - Sample`. You should see the following text logged to your `Debug Console`:
```
Reading directory /
Creating directory /aaa
Reading directory /aaa
Reading directory /
Reading directory /aaa
```

I would expect to see the same behavior in both cases. Discovered during the investigation of https://github.com/intersystems-community/vscode-objectscript/issues/809

VS Code version: Code 1.63.2 (899d46d82c4c95423fb7e10e68eba52050e30ba3, 2021-12-15T09:37:28.172Z)
OS version: Darwin x64 21.2.0
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz (12 x 2900)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|32.00GB (7.65GB free)|
|Process Argv|--enable-proposed-api intersystems-community.vscode-objectscript --crash-reporter-id 8728ef25-68c4-4e80-aafe-921e2dc7ba3b|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (4)</summary>

Extension|Author (truncated)|Version
---|---|---
servermanager|int|2.0.6
vscode-objectscript|int|1.2.3-dev.807
language-server|int|1.2.7
vscode-docker|ms-|1.18.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30410667
vscop804:30404766
vs360:30404995
vsrem710cf:30405999
vscexrecpromp3c:30407761

```

</details>

**Comment by bpasero** [comment]
Confirmed this running locally. I can see a `IFileService.readdir` call here:

https://github.com/microsoft/vscode/blob/0cfbbaeda89cdc1955af32cce608252810c35c20/src/vs/platform/files/common/fileService.ts#L256

Triggered by `explorerModel` it seems:

```
toFileStat (fileService.js:204)
doResolveFile (fileService.js:166)
processTicksAndRejections (internal/process/task_queues.js:93)
await in processTicksAndRejections (async)
resolve (fileService.js:147)
fetchChildren (explorerModel.js:234)
getChildren (explorerViewer.js:47)
(anonymous) (asyncDataTree.js:581)
createCancelablePromise (async.js:15)
doGetChildren (asyncDataTree.js:580)
doRefreshNode (asyncDataTree.js:552)
doRefreshSubTree (asyncDataTree.js:532)
(anonymous) (asyncDataTree.js:534)
doRefreshSubTree (asyncDataTree.js:534)
processTicksAndRejections (internal/process/task_queues.js:93)
await in processTicksAndRejections (async)
refreshNode (asyncDataTree.js:521)
refreshAndRenderNode (asyncDataTree.js:508)
_updateChildren (asyncDataTree.js:355)
updateChildren (asyncDataTree.js:344)
refresh (explorerView.js:432)
refresh (explorerService.js:222)
(anonymous) (explorerService.js:103)
fire (event.js:464)
(anonymous) (event.js:54)
(anonymous) (event.js:65)
fire (event.js:464)
(anonymous) (event.js:43)
fire (event.js:464)
(anonymous) (event.js:54)
fire (event.js:464)
handler (ipc.js:472)
onResponse (ipc.js:541)
onBuffer (ipc.js:530)
(anonymous) (ipc.js:364)
fire (event.js:464)
fn (event.js:261)
emit (events.js:315)
onMessage (electron/js2c/renderer_init.js:93)
toFileStat (fileService.js:204)
doResolveFile (fileService.js:166)
processTicksAndRejections (internal/process/task_queues.js:93)
await in processTicksAndRejections (async)
resolve (fileService.js:147)
createFolder (fileService.js:707)
processTicksAndRejections (internal/process/task_queues.js:93)
await in processTicksAndRejections (async)
(anonymous) (workingCopyFileService.js:89)
doCreateFileOrFolder (workingCopyFileService.js:89)
processTicksAndRejections (internal/process/task_queues.js:93)
await in processTicksAndRejections (async)
createFolder (workingCopyFileService.js:62)
perform (bulkFileEdits.js:166)
apply (bulkFileEdits.js:341)
_performFileEdits (bulkEditService.js:101)
perform (bulkEditService.js:84)
apply (bulkEditService.js:179)
(anonymous) (explorerService.js:131)
(anonymous) (progressService.js:153)
withNotificationProgress (progressService.js:144)
withProgress (progressService.js:59)
applyBulkEdit (explorerService.js:125)
onSuccess (fileActions.js:700)
onFinish (fileActions.js:724)
await in onFinish (async)
(anonymous) (explorerViewer.js:307)
(anonymous) (functional.js:18)
(anonymous) (explorerViewer.js:334)
(anonymous) (dom.js:52)
```

**Comment by omerfarman2** [comment]
``

**Comment by isc-bsaviano** [comment]
@JacksonKearl Now that 1.66 is out, what are the chances that this will be worked on in the current iteration?

**Comment by JacksonKearl** [comment]
Could you help me understand the user facing problem here? Ideally reading the directory would be idempotent, we can maybe get the explorer not to here but extensions and similar can read the dir whenever they want.

**Comment by JacksonKearl** [comment]
Digging into this the root cause is that when we try to get the workspace folder of a resource we use effectively a prefix query, finding the workspace folder that matches the given uri with the longest prefix. However this breaks down in the case of query parameters, where `memfs:/?notaparam=1` is not a prefix of `memfs:/aaa?notaparam=1`. 

The fix for this is not obvious to me... we could change the workspace folder finder to treat query parameters as higher "priority" than paths, so internally the comparison would look more like `memfs:?notaparam=1/` vs. `memfs:?notaparam=1/aaa`, but I'm not confident that wont break something else.

I have potential fix up on the branch with [c759f6b](https://github.com/microsoft/vscode/commit/c759f6b669777d3fc0e3b1c8378ad461d2b19889), but again it's a very core routine and I'm not sure this change is correct.

**Comment by isc-bsaviano** [comment]
/verified

Thanks @JacksonKearl! This looks good in the Insiders build for this morning.

**Comment by isc-kiyer** [comment]
I am still able to reproduce this in the following VS code version:

Version: 1.67.1 (Universal)
Commit: da15b6fd3ef856477bf6f4fb29ba1b7af717770d
Date: 2022-05-06T12:37:16.526Z
Electron: 17.4.1
Chromium: 98.0.4758.141
Node.js: 16.13.0
V8: 9.8.177.13-electron.0
OS: Darwin x64 20.6.0

Did this change not make it into the above version?

cc @isc-bsaviano

**Comment by JacksonKearl** [comment]
@isc-kiyer I would expect the fix to be in 1.67.1
